mods/tnt/init.lua

```diff
@@ -85,27 +82,26 @@ local function add_drop(drops, item)
 	end
 end
 
-local fire_node = {name="fire:basic_flame"}
 
-local function destroy(drops, pos, cid)
-	if minetest.is_protected(pos, "") then
-		return
+local function destroy(drops, npos, cid, c_air, c_fire, on_blast_queue, ignore_protection, ignore_on_blast)
+	if not ignore_protection and minetest.is_protected(npos, "") then
+		return cid
 	end
 	local def = cid_data[cid]
-	if def and def.on_blast then
-		def.on_blast(vector.new(pos), 1)
-		return
+	if not ignore_on_blast and def and def.on_blast then
+		on_blast_queue[#on_blast_queue + 1] = {pos = vector.new(npos), on_blast = def.on_blast}
+		return cid
 	end
 	if def and def.flammable then
-		minetest.set_node(pos, fire_node)
+		return c_fire
 	else
-		minetest.remove_node(pos)
 		if def then
 			local node_drops = minetest.get_node_drops(def.name, "")
 			for _, item in ipairs(node_drops) do
 				add_drop(drops, item)
 			end
 		end
+		return c_air
 	end
```

Comment:
This could be restructured a bit to reduce indentation and only check `def` once:
```Lua
if not def then
	return c_air
elif def.flamable then
	return c_fire
else
	...
end
```